README.md
This repo contains:

- source files for  'Use R, Canada!' Portal (<https://open-canada.github.io/UseR/>): located here: <https://github.com/open-canada/UseR/tree/gh-pages>
- open source R codes used in GC Lunch and Learn R Friday's weekly meetups (TBA)




IVIM/NEWS.md
# IVIM 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


IVIM/iviDT.R

# IVIM/iviBase.R ----

# I put all imported package on top

#' Generic utility functions to work with IVIM package
#'
#' @param x 1st
#' @param y 2nd
#'
#' @import data.table
#' @import magrittr
# @import ggplot2
# @import lubridate


#' @return 1st2nd
#' @rdname iviBase
#' @export
`%+%` <- function (x,y) {
  paste0(x,y)
}

#' @return Return x without y
#' @rdname iviBase
#' @export
`%wo%` <- function(x, y) {
  x[!x %in% y]
}  # names(iris) %wo% "Species"



#' @return Returns TRUE  if x that are NOT in y
#' @rdname iviBase
#' @export
'%notin%' <- function (x,y) {  ! ( x %in% y)   }

'%ni%' <- '%notin%'
# `%ni%` <- Negate(`%in%`) #names(iris) %notin% "Species"



# these are added above using #' @import ... commands
# Se still have this function - for testing here and callin from outside
#' Title
#'
#' @return opens packages used by IVIM package
#' @export
#'
#' @examples
iviPackages <- function() {
  library(magrittr)
  library(data.table);
  # library(dplyr) # can I avoid using dplyr altogether,if I use data.table?
  library(lubridate);
  library(stringr)
}



#' Set options for printing and calling data
#'
#' @param rownames Boolean - show row number?
#' @param topn number of top/bottom rows to show
#' @param nrows rows to print
#' @param digits digits to print
#'
#' @return
#' @rdname iviBase
#' @export
#'
#' @examples iviOptions()
iviOptions <- function(rownames=T, topn=5, nrows = 50, digits = 3) {

  # options() # shows all options available
  # options(defaultPackages = c("datasets",  "iviDT" ,    "grDevices", "graphics",  "stats",     "methods") )
  options(digits = digits)  # 7
  # options(max.print = 100) # 1000
  options(scipen = 999)  # remove scientific notation
  # options(na.action = na.omit) #, na.fail, also for na.exclude, na.pass, na.omit
  # getOption("na.action")

  options(lubridate.week.start = 1) # 7

  options(datatable.verbose = FALSE) # TRUE)
  options(datatable.print.class = TRUE)
  options(datatable.print.keys = TRUE)
  options(datatable.print.nrows = nrows) # 100
  options(datatable.print.topn=topn) # 5
  options(datatable.print.rownames = rownames)
}

# IVIM/iviDT.R  ----

#' Convenience functions from IVIM package to work with data.table object.
#' #' In place (i.e., by memory pointer) value replacement in a data.table column
#' @param dt A datatable object
#' @param col The column of interest
#' @param a The value you wish to replace
#' @param b The replacement for value a
#'
#' @return This function modifies data.table dt by memory reference (i.e, in place, without creating new objects). Hence, it does not need to return anything. However when desired (e.g., for piping with '%>%'), it will return the data.table dt that was modified.
#
# This very incovenient to write R code example as text comment ! -
# DISCUSS IT! - It was suggested to put any such R codes in a) MY_CODE/test-*.R  and b) in tests/testthat/test-DT.R
#
#' @examples \dontrun{
#' province <- c("Alberta", "Quebec")
#' dt <- data.table(province)
#' dt %>% dt.replaceAwithB("province", "Alberta", "AB")
#' }
#' @rdname iviDT
#' @export
dt.replaceAwithB <- function(dt, col, a, b) {
  dt[get(col)==a, (col):=b];
}

# In place (i.e., by memory pointer) value replacement in a data.table column
#
# @param dt
#  @param col
# @param a
# @param b
#
#' @return This function performs the inverse operation to `dt.replaceAwithB()` function: i.e. `dt %>% replaceAwithB(col, a, b) %>% dt.replaceBwithA(col, a, b) == dt` # TRUE
#' @rdname iviDT
#' @export
dt.replaceBwithA <- function(dt, col, a, b) {
  dt[get(col)==b, (col):=a];
}

if (F) {
  # This should be Moved to vignette !

  # TEST 1: ways to pipe  ----

  library(IVIM)
  iviPackages()
  # library(data.table)
  # library(magrittr)

  dt1 <- data.table(province=c("Alberta", "Quebec"))

  dt1 %>%
    dt.replaceAwithB ("province", "Quebec", "QB") %>%
    dt.replaceAwithB ("province", "Alberta", "AB") %>%
    dt.replaceAwithB ("province", "Alberta", "AB");
  dt1
  # bug here ? - it does not print dt1 ! -  you need to call dt1 again, or call dt1 %>% print
  # ask at stackoverflow !
  dt1

  # reverse the changes:
  dt1 %>%
    dt.replaceBwithA ("province", "Quebec", "QB") %>%
    dt.replaceBwithA ("province", "Alberta", "AB") %>%
    dt.replaceBwithA ("province", "Alberta", "AB") %>% print

  if (F) {
    # NOte: you can pipe the same functionality without dt.replaceAwithB function

    # Way 1: using data.table [][][] operator

    # I prefer this style (so that to have [] in final line):
    dt1[province == "Quebec", province:="QB"
        ][province == "Alberta", province:="AB"
          ][province == "Alberta", province:="AB"
            ][]
    # Alternative style:
    # dt1[province == "Quebec", province:="QB"][
    #   province == "Alberta", province:="AB"][
    #     province == "Alberta", province:="AB"][]
    dt1

    # Now the same using dt.replaceAwithB function

    # Way 2: using magrittr %>% operator
    # If you update to R 4.1.0, there is a base pipe operator now, |>, which may get rid of any need to use magrittr.

    dt1[province== "Quebec", province:="QB"] %>%
      .[province== "Alberta", province:="AB"] %>%
      .[province== "Alberta", province:="AB"] %>% .[]

  }

  # TEST 2: comparisons w. iris----

  # Standard way (least efficient - using dplyr w/o data.table)
  df <- iris[100:101,]; df

  # R base:
  df0 <- df
  df0$Species[df$Species == 'virginica'] <- 'virgin.' #bad
  df; df0

  # library(dtplyr)
  df0 <- df  %>%  dplyr::filter(Species=="virginica") %>% dplyr::mutate(Species="virgin.") #bad
  df; df0
  df0 <- df %>% dplyr::mutate(Species=ifelse(Species=="virginica", "virgin.", Species)) #bad
  df; df0
  df0 <- df %>% dplyr::mutate(Species = forcats::fct_recode(Species, "virgin." = "virginica" ))  #works
  df; df0


  # Compare to (not much better - using Wickam's dtplyr way around with data.table):
  library(dtplyr)
  dtBackEnd <- lazy_dt(df)
  dtBackEnd0 <- dtBackEnd %>%
    filter(Species=="virginica") %>%
    mutate(Species="virgin.") %T>%
    show_query();
  dtBackEnd0;
  dtBackEnd;
  dtBackEnd0 %>% as.data.table()

  #  Compare to (the best way):
  dt <- copy(df);  # dt is data.frame
  dt %>% setDT; # NB: dt is now data.table and you can use it with dt.*() functions !
  dt %>%
    dt.replaceAwithB("Species", "virginica", "virgin.") %>%
    dt.replaceAwithB("Species", "versicolor", "versic.");
  dt; df # NB: dt also changed !

}



# replace set A with set B in column col

# Very useful for merging PSES responses from various years, because they different numbering!

# Asked: https://stackoverflow.com/questions/67608799/function-to-replace-values-in-data-table-using-a-lookup-table

#' Title
#'
#' @param dt
#' @param col
#' @param dtLookup
#' @rdname iviDT
#' @export

dt.replaceValueUsingLookup <- function(dt, col, dtLookup) {
  dt[
    dtLookup,
    on = setNames("old", col),
    (col) := new
    ]
}
if (F) {
  dt <- data.table( chapter=as.character(11:15) );dt

  dtLookup <- data.table(
    old = c("11", "12", "14", "15"),
    new = c("101", "102", "105", "104")
  )
  dt %>% dt.replaceValueUsingLookup("chapter", dtLookup)
}




#' @rdname iviDT
#' @export
dt.reset <- function(dt) {
  setDT(setDF(dt))
  # can we do it w/o assignment ? (by pointers?)
  # dt <- setDT(setDF(dt))
}
if (F) { # TEST 1 ----
  dtProjectile <- data.table()
  dtProjectile$angle<- c(15, 30, 45, 60, 75)
  dtProjectile$distance <- c(5.1, 8, 10,  8.5, 4.8)
  dtProjectile <- dt.reset(dtProjectile) # setDT(setDF(dtProjectile))
  ggplot() + geom_point( aes(dtProjectile$angle, projectile$distance)  ) # Works
  ggplot(dtProjectile) + geom_point( aes(angle,distance) ) # Does not work UNLESS setDT(setDF())
}

#' Keep or remove columns in data.table by memory reference
#' @rdname iviDT
#' @examples
#' dt<-iris; dt %>%  setDT %>% dt.rmCols (1) %>% dt.rmCols("Species"); dt
#' @export
dt.rmCols <- function (dt, cols){
  dt [, (cols):=NULL]
}

#' @rdname iviDT
#' @export
dt.keepCols <- function (dt, cols) {
  if (is.integer(cols))
    dt.rmCols(dt, 1:ncol(dt) %wo% cols)
  else
    dt.rmCols(dt, names(dtCases) %wo% cols)
}





packages101/package_initialization_script.R
library(devtools)

devtools::create("rCanada") # if you are  in /my_package
devtools::create("../../rCanada") # will it erase it,if I lready have it?
> #or usethis::create_package("../r4gc/packages/rCanada2")

library(roxygen2)
# library(testthis). What's the difference from library(testthat) ??
library(testthat) # do we need it?
# Attaching package: ‘testthat’
#
# The following object is masked from ‘package:devtools’:
#
#   test_file
#
# The following objects are masked from ‘package:magrittr’:
#
#   equals, is_less_than, not


usethis::use_testthat() # from usethis, which is callsed from devtools


use_vignette(name="my_vignettes") #
use_vignette(name="data_linking")


use_package("data.table")
use_package("magrittr")
use_package("data.table")
use_package("lubridate", type="Imports")  # how about order ?! lubridate must be after data.table !
use_package("stringr")



IVIM/utils-MASTER_COPY (under development).R
# MASTER COPY for functions used in utils.R in other packages

# utils.R
#
# Author: Dmitry Gorodnichy
# Tested, Can be run as source("utils.R")


### For IVIM package. ----

# Needed (for THIS package) functions/codes below are extracted from master copy:
# my_packages/IVIM/CODES/*
# also, https://gccode.ssc-spc.gc.ca/gorodnichy/IVIM/CODES
# also ??, https://gccode.ssc-spc.gc.ca/gorodnichy/rCanada/R/utils.R

# These are placed  in utils.R in all packages that  I'm developing until IVIM package is on CRAN.
# Once it's on CRAN, I'll be able to import it as data.table or any other package.

# imported libraries ----

setPackages <- function() {

  library(magrittr)
  library(data.table);
  options(datatable.print.class=TRUE) # NB: library(data.table) MUST BE BEFORE lubridate !!
  # library(dplyr) # can I avoid using dplyr altogether,if I use data.table?
  library(lubridate);
  options(lubridate.week.start =  1)
  library(stringr)
}

"%+%" <- function (x,y) paste0(x,y)
"%wo%" <- function(x, y) x[!x %in% y]  # names(iris) %wo% "Species"
`%ni%` = Negate(`%in%`) # x not in y # names(iris) %in% "Species";  names(iris) %ni% "Species"
'%notin%' = Negate(`%in%`) #names(iris) %notin% "Species"


# IVIM/iviBase.R ----
# (maybe aso iviPlot.R)

setMyOptions <- function() {
  options() # shows all options available
  # options(defaultPackages = c("datasets",  "utils" ,    "grDevices", "graphics",  "stats",     "methods") )
  options(digits = 3)  # 7
  # options(max.print = 100) # 1000
  options(scipen = 999)  # remove scientific notation
  # options(na.action = na.omit) #, na.fail, also for na.exclude, na.pass, na.omit
  # getOption("na.action")

  options(lubridate.week.start = 1) # 7

  dt.setOptions()
}

# . GLOBALS ----

iviGLOBALS <- list(
  VERBOSE = TRUE,
  COMMENT_OUT = F,
  USE_PNG = F,
  COLOURS = c("black", "green", "red", "blue", "brown", "cyan", "violet", "orange"),
  # COLOURS=rainbow(14),  # COLOURS=colours()[round(seq(1,length(colours()),length.out=11))],
  COUNT = 1,
  resetCOUNT = function() iviGLOBALS$resetCOUNT <- 1
) #

# Function to specify a series of numbers that loop around
loop <- function(...) {
  numbersToLoopAround <- c(...)
  i <- 0
  function() {
    if (i < length(numbersToLoopAround)) i <<- i + 1  else    i <<- 1
    return(numbersToLoopAround[i])
  }
}
if (F) {
  setNewColour <- loop(iviGLOBALS$COLOURS)
  for (i in 1:10) print(setNewColour())
}



# Asked and answered at Stackoverflow
exit <- function() {
  .Internal(.invokeRestart(list(NULL, NULL), NULL))
  # All below don't work:
  # q()
  # stop("Stopping");  # This generated "Error in " and goes to Debug mode, which we dont want.
  # return (invisible());
}

dd.getKey <- function(prompt = "-", lastkey = "y") {
  if (tolower(lastkey) == "a") {
    return("a")
  }
  strKey <- readline(paste(prompt))
  nKeys <- as.integer(charToRaw(strKey)) - as.integer(charToRaw("0"))
  if (length(nKeys) == 0) {
    return("y")
  } else if (nKeys[1] >= 0 & nKeys[1] <= 9) {
    return(nKeys)
  } else if (strKey == "q") {
    exit()
  } else {
    return(strKey)
  }
}

dd.ConfirmedYes <- function(...) {
  if (dd.getKey(...) %in% c("y", "a")) {
    return(TRUE)
  } else {
    return(FALSE)
  }
}

# text view of data.table on screen

dt.iPrint <- function(dt, howmuch = 22) {

  cKey <- dd.getKey("Start printing all lines ?  ")
  incr <- 25
  n <- 1
  while (n < nrow(dt)) {
    print(dt[n:(n + incr - 1)])
    n <- n + incr
    print(n)
    cKey <- dd.getKey("Continue displaying data ? (y/n/1,2,3 to increase number of printed lines) ")
    if (is.integer(cKey)) {
      incr <- (cKey[1] + 1) * 25
    } else if (tolower(cKey) == "n") break
  }
}


if (F) {
  dt <- as.data.table(iris)
  dt %>% dt.iPrint()
}

# IVIM/iviDT.R (aka dt-helper.R) ----

# See also:

# - https://github.com/tidyverse/dtplyr (Version: 1.1.0, Published: 2021-02-20, From Hadley himself - I found it quite cumbersome still though...)
# https://github.com/asardaes/table.express (Version: 0.3.1 Published: 2019-09-07 - somewhat easier?)
# https://github.com/asardaes/table.express/blob/master/R/pkg.R
# - https://github.com/markfairbanks/tidytable (Version: 0.6.2 Published: 2021-05-18 - seems to be the best supported of the three?)



dt.setOptions <- function(rownames=T, topn=5, nrows = 50) {
  options(datatable.verbose = FALSE) # TRUE)
  options(datatable.print.class = TRUE)
  options(datatable.print.keys = TRUE)
  options(datatable.print.nrows = nrows) # 100
  options(datatable.print.topn=topn) # 5
  options(datatable.print.rownames = rownames)
}

# sometimes (eg RecordLinkage::names ) you can't use setDT() and have to use as.data.table()
# Error in setDT(iris) :
#   Cannot convert 'iris' to data.table by reference because binding is locked. It is very likely that 'iris' resides within a package (or an environment) that is locked to prevent modifying its variable bindings. Try copying the object to your current environment, ex: var <- copy(var) and then using setDT again.
# dt.set2 <- function(...) {
#   as.data.table(...) # or setDT(), or as.data.table(), or data.table
# }
# if (F) {
#
# }

# to be able to update data.table on the fly: from empty DT to non-empty DT
dt.set <- function(dt) {
  setDT(setDF(dt))
}
if(F) {
  library(data.table); library(ggplot2)
  dtProjectile <- data.table()
  dtProjectile$angle<- c(15, 30, 45, 60, 75)
  dtProjectile$distance <- c(5.1, 8, 10,  8.5, 4.8)

  ggplot() + geom_point( aes(dtProjectile$angle, dtProjectile$distance)  ) # always Works
  ggplot(dtProjectile) + geom_point( aes(angle,distance) ) # works only if you run dt.set(dtProjectile))!

  ggplot(dt.set(dtProjectile)) + geom_point( aes(angle,distance) )
  # So don't forget to run  it:
  # dt.set(dtProjectile)
}




# These are better (ie in-place, by pointers) alternatives to those in dplyr or dtply

# Instead of: dt <-  dt %>% dtply::select(-cols); dt <-  dt %>% dtply::select(cols);

dt.rmCols <- function (dt, cols){
  dt [, (cols):=NULL]
}
dt.keepCols <- function (dt, cols) {
  if (is.integer(cols))
    dt.rmCols(dt, 1:ncol(dt) %wo% cols)
  else
    dt.rmCols(dt, names(dtCases) %wo% cols)
}

# replace A with B in column col

dt.replaceAwithB <- function(dt, col, a, b) {
  dt[get(col)==a, (col):=b]; return(dt)
}

# . TEST ME ----
if (F) {
  df <- iris;
  dt <- lazy_dt(df)
  dt3 <- dt %>%
    filter(Species=="virginica") %>%
    mutate(Species="virgin.") %>%
    show_query()

  dt2 <- dt1 <- df %>% setDT
  dt1[Species=="virginica", Species:="virgin."]
  dt2 %>% dt.replaceAwithB(Species, "virginica", "virgin.")
}



# replace set A with set B in column col

# Very useful for merging PSES responses from various years, because they different numbering!

# Asked: https://stackoverflow.com/questions/67608799/function-to-replace-values-in-data-table-using-a-lookup-table

dt.replaceValueUsingLookup <- function(dt, col, dtLookup) {
  dt[
    dtLookup,
    on = setNames("old", col),
    (col) := new
    ]
}
if (F) {
  dt <- data.table( chapter=as.character(11:15) );dt

  dtLookup <- data.table(
    old = c("11", "12", "14", "15"),
    new = c("101", "102", "105", "104")
  )
  dt %>% dt.replaceValueUsingLookup("chapter", dtLookup)
}





dt.apply <- function(dt, functionToApply, cols) {
  if (cols %>% is.integer() )
    cols <- names(dt)[column]
  dt[, (cols) := lapply(.SD, functionToApply),.SDcols = cols]
}
if (F) {
  dt <- iris %>% setDT; dt[1]
  column <- "Species"
  funcToDo <- as.character
  dt %>% data.table() %>% dt.apply(as.ordered,"class_id") %>% summary
}





# NOT SURE how it works, but it works!  Found in: stackoverflow:
# Extremely fast Efficient rows delition from data.table

# dt.rows_remove
dt.rmRows <- function(DT, del.idxs) {  # pls note 'del.idxs' vs. 'keep.idxs'
  if (!is.data.table(dt))
    dt <- as.data.table(dt)

  keep.idxs <- setdiff(DT[, .I], del.idxs);  # select row indexes to keep
  cols = names(DT);
  DT.subset <- data.table(DT[[1]][keep.idxs]); # this is the subsetted table
  setnames(DT.subset, cols[1]);
  for (col in cols[2:length(cols)]) {
    DT.subset[, (col) := DT[[col]][keep.idxs]];
    DT[, (col) := NULL];  # delete
  }
  return(DT.subset); # NB: Original DT is also changed  by reference !
}


# Extremely fast subset function
# dt[, .SD[1:10], by=y] # too slow
# dt %>% top_n(2, y) # even slower

dt.subset <- function (dt, by_variable, subset_size) {
  if (!is.data.table(dt))
    dt <- as.data.table(dt)
  dt <- dt[dt[, .I[1:subset_size], keyby = by_variable]$V1]
}
if (F) {
  dt <- attrition; by_variable <- "Attrition"; subset_size <- 10
  dt <- attrition %>% dt.subset("Attrition",5)
}


# --------------------------------------------------------------------- -#
# IVIM/iviReportivi.R ----
# --------------------------------------------------------------------- -#


rmd.cat <- function(...) {
  cat("\n\n");   cat (...);   cat("\n\n")
}
rmd.print <- function(...) {
  cat("<pre>\n")
  print (...)
  cat("\n</pre>")
}

set.seed(22)  #btn-block

DIV_START <- function(strText="Details", bShow=T){
  iButton <- runif(1, min=0, max=100000) %>% as.integer()
  # "btn btn-primary btn-block btn-lg active"
  rmd.cat( glue::glue(  paste0(
    '<button class="btn btn-primary btn-block active"  data-toggle="collapse" data-target="#BlockName',iButton, '"> Show/Hide ', strText, '</button>',   '\n',
    '<div id="BlockName', iButton,'" class="collapse', ifelse(bShow == F, "", " in"), '">'
  )       )       )
}
DIV_END <- function(str="") {
  rmd.cat("</div>")
}

if (F) {
  cat ( DIV_START("Details", bShow=F) )
  for (i in 1:nrow(mtcars)) cat (paste(mtcars[i,],collapse = " " ), "\n")
  cat (DIV_END())
}

d7.datatable <- function(x, rownames=T){
  DT::datatable(x,
                rownames=rownames,
                filter="top",
                extensions = 'Buttons',
                options = list(dom = 'Blfrtip',
                               buttons = c('copy', 'csv', 'excel', 'pdf', 'print'),
                               lengthMenu = list(c(5,10,25,-1),
                                                 c(5,10,25,"All"))))
}



# IVIM/iviPlot.R ----


# dd.plotDtAllBoxplots(dt, c("BWT_traveller", "HH", "OFFICE"), scales="free")

#up to five variables can be plotted
d7.plotAllBoxplots <- function (dt, columns, scales="free_x"){
  columns = dd.names(dt, columns)

  plot = ggplot(dt) + labs( y=columns[1], x=columns[2],
                            title=sprintf("%s = f(%s) by (%s,%s,%s))",columns[1],columns[2],columns[3],columns[4],columns[5]) )

  if (length(columns) < 5) {
    plot = plot + geom_boxplot( aes(x = as.ordered(get(columns[2])), y = get(columns[1])), varwidth=T) #, notch=T )
  }
  else {
    plot = plot + geom_boxplot( aes(x = as.ordered(get(columns[2])), y = get(columns[1])), varwidth=T )
    #                                 position=as.ordered(as.name(columns[5])), varwidth=T )
    # Error in get(columns[5]) : object 'YY' not found
  }

  if (length(columns) == 2)      { plot }
  else if (length(columns) == 3) { plot + facet_wrap(~as.ordered(get(columns[3])), scales=scales) } # scales="free", "free_y"
  else if (length(columns) == 4) { plot + facet_grid(as.ordered(get(columns[4])) ~ as.ordered(get(columns[3])), scales=scales) }
  else if (length(columns) > 5)  { print (paste("Opps - incorrect number of  columns: ", length(columns))) }
}


d7.plotAllHist <- function(dt) {
  #aa.dtPlotAllHist <- function(dt) {
  parOld <- par(mfrow = c(2, 1))
  for (col in 1:ncol(dt)) {
    colName <- names(dt)[col]
    x <- dt[[colName]]

    if ( !is.numeric( x ) ) {
      dt[[colName]] <- as.ordered( x )
      cFivenum <- sprintf("   Factor with %i labels: %s, %s, ... ",
                          length( labels(summary(x))),
                          labels(summary(x))[1] ,labels(summary(x))[2])
    }
    else{
      #    a <- quantile( x ) # fivenum(x); summary(x)
      #    cFivenum <- sprintf("    Quantiles: %s, %s, %s, %s, %s", a[1],a[2],a[3],a[4],a[5])
      a <- summary(x)
      cFivenum <- sprintf("    Quantiles: %s, %s, %s, %s, %s (NA=%s)", a[[1]],a[[2]],a[[3]],a[[4]],a[[5]], a[[7]])
    }

    print( sprintf("Column: #%i. column name: `%s` ", col, colName ) )
    print( cFivenum )
    plot( x, main=sprintf("Column %i: `%s`",col, colName), xlab=cFivenum)
    hist(as.numeric(x), main="", xlab="") #, breaks = 100) #labels = TRUE, col = "gray",
  }

  par(parOld)
}



# IVIM/iviLinking.R ----

library(RecordLinkage)
# Loading required package: DBI
# Loading required package: RSQLite
# Loading required package: ff
# Loading required package: bit
#
# Attaching package: ‘bit’
#
# The following object is masked from ‘package:base’:
#
#   xor
#
# Attaching package ff
# - getOption("fftempdir")=="C:/Users/gxd006/AppData/Local/Temp/Rtmp23BN4P/ff"
#
# - getOption("ffextension")=="ff"
#
# - getOption("ffdrop")==TRUE
#
# - getOption("fffinonexit")==TRUE
#
# - getOption("ffpagesize")==65536
#
# - getOption("ffcaching")=="mmnoflush"  -- consider "ffeachflush" if your system stalls on large writes
#
# - getOption("ffbatchbytes")==184467440737095520 -- consider a different value for tuning your system
#
# - getOption("ffmaxbytes")==9223372036854775808 -- consider a different value for tuning your system
#
#
# Attaching package: ‘ff’
#
# The following objects are masked from ‘package:utils’:
#
#   write.csv, write.csv2
#
# The following objects are masked from ‘package:base’:
#
#   is.factor, is.ordered
#
# RecordLinkage library
# [c] IMBEI Mainz
#
#
# Attaching package: ‘RecordLinkage’
#
# The following object is masked from ‘package:bit’:
#
#   clone
#
# The following object is masked from ‘package:base’:
#
#   isFALSE


